### PR TITLE
NullVm: fix bugs when linking with Rust SDK.

### PR DESCRIPTION
**Pull request by mathetake**
I have been playing around with Rust SDK in my spare time and [statically linking plugins in Rust with Envoy](https://github.com/mathetake/proxy-wasm-rust-nullvm). I found that some changes must be made in the cpp-host to resolve the Rust SDK specific issues:
- ~need to set `Cloneable:CompiledBytecode` for Rust since we need to call `_start` on all threads.~ fixed in the Rust Null VM SDK(https://github.com/mathetake/proxy-wasm-rust-nullvm/commit/ddfe09fb8914f93acc2439f00a6e486cc90856bc)
- need to clean up root contexts created during configuration_canary calls since the code panics due to the duplicated on_create_context call for the same root context id (=1).

This is still WIP since I haven't figured out how to fix Envoy tests, but all of them are uninteresting mock function calls and can be easily fixed.

With this change, it looks like we can link Rust SDK with Envoy without modifying Rust SDK itself and run them as NullVms without issues. I think this is awesome!

FYI, we can link multiple plugins at the same time without symbol collision with [my simple trick here](https://github.com/mathetake/proxy-wasm-rust-nullvm/blob/main/bazel/proxy_wasm_symbols.bzl#L58-L68) :slightly_smiling_face: 

ref: 
- https://github.com/envoyproxy/envoy/issues/12155 
- https://github.com/mathetake/proxy-wasm-rust-nullvm

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: include/proxy-wasm/wasm.h [hunk 1]
```diff
@@ -290,7 +290,11 @@ class WasmHandleBase : public std::enable_shared_from_this<WasmHandleBase> {
     }
   }
 
-  void kill() { wasm_base_ = nullptr; }
+  void kill() {
+    wasm_base_->startShutdown();
+    wasm_base_->finishShutdown();
+    wasm_base_ = nullptr;
```
Comment (PiotrSikora) [inline include/proxy-wasm/wasm.h:296]: Calling `finishShutdown()` immediatelly after `startShutdown()`, without waiting for the plugin to call `proxy_done` is technically incorrect, and effectlively a no-op. This sequence works only if the plugin returns `true` from the `proxy_on_done` callback, in which case you don't even need to call `finishShutdown()`.

It also breaks the promise of not executing anything in the canary (e.g. you can imagine plugins sending batch reports on shutdown, etc.), which is why we're simply dropping the VM right now.

File: src/null/null_vm.cc [hunk 2]
```diff
@@ -36,6 +36,13 @@ RegisterNullVmPluginFactory::RegisterNullVmPluginFactory(std::string_view name,
   (*null_vm_plugin_factories_)[std::string(name)] = factory;
 }
 
+Cloneable NullVm::cloneable() {
+  if (!plugin_) {
+    return Cloneable::NotCloneable;
+  }
+  return plugin_->cloneable();
+}
```
Comment (PiotrSikora) [inline src/null/null_vm.cc:44]: `Cloneable` is a property of a runtime, so it shouldn't change based on whether or not the plugin was already loaded. It seems like a workaround for an issue elsewhere.

**Review by PiotrSikora — COMMENTED** [review]
> With this change, it looks like we can link Rust SDK with Envoy without modifying Rust SDK itself and run them as NullVms without issues. I think this is awesome!

This is indeed awesome! Thanks for all this work, this is great!

> * need to set `Cloneable:CompiledBytecode` for Rust since we need to call `_start` on all threads.

You mention `Cloneable::CompiledBytecode`, but the only `Cloneable` values in this PR are `Cloneable::NotCloneable` and `Cloneable::InstantiatedModule`.

> * need to clean up root contexts created during configuration_canary calls since the code panics due to the duplicated on_create_context call for the same root context id (=1).

What's the difference between plugins written using C++ SDK and Rust SDK here? Do C++ plugins re-use the root context created during canary? That would be wrong.

Basically, I'm trying to understand why this isn't the issue for the existing NullVM C++ plugins.

File: src/null/null_vm.cc (hunk 2 above)
Comment (mathetake) [reply src/null/null_vm.cc:44]: agreed. Le me find more proper place to fix.

**Comment by mathetake** [comment]
> You mention Cloneable::CompiledBytecode, but the only Cloneable values in this PR are Cloneable::NotCloneable and Cloneable::InstantiatedModule.

Yes, I had an intention to implement `cloneable()` in Rust's NullVm here: https://github.com/mathetake/proxy-wasm-rust-nullvm/blob/main/include/proxy-wasm-rust-nullvm/rust_nullvm.h#L272

> Basically, I'm trying to understand why this isn't the issue for the existing NullVM C++ plugins.

Yeah, actually I was also confused by why C++ works so far. But I notice that [C++ SDK DOES allow on_context_create calls for same context ids](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/master/proxy_wasm_intrinsics.cc#L60-L63). In contrast, [Rust SDK (and Go SDK as well) panics](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/master/src/dispatcher.rs#L105) in that case. This is the why I need to clean up contexts created in the configuration canary phase for Rust SDK.

BTW, where do you think is appropriate to put my Rust SDK for NullVm(https://github.com/mathetake/proxy-wasm-rust-nullvm/blob/main/include/proxy-wasm-rust-nullvm/rust_nullvm.h)? Here, or proxy-wasm-rust-sdk repo? NullVm is cpp-host specific stuff but the glue code itself contains Rust specific code, so either seems fine to me.

File: src/null/null_vm.cc (hunk 2 above)
Comment (mathetake) [reply src/null/null_vm.cc]: https://github.com/mathetake/proxy-wasm-rust-nullvm/commit/ddfe09fb8914f93acc2439f00a6e486cc90856bc can be fixed in the glue code.

**Comment by mathetake** [comment]
So basically, including C++ null plugins, on_vm_start and on_configure are called twice ( canary + the usual startup time call). I think we should fix this not only for Rust but also C++.

**Comment by mathetake** [comment]
this should be fixed when we removing the need for canary. Closing.

**Pull request closed without merging**